Use julia-repl in manual & other edits

mortenpi opened the pull request:
Use `julia-repl` to highlight REPL blocks in the manual, which makes them consistent with doctest blocks. Requires Documenter v0.10.1, which is why it's currently based on #21846.

I also came across a number of REPL blocks that seemed like they could be turned into doctests as well, and a few other small things. However, I guess there might have been a good reason why some of them were not enabled, so I'd be happy to drop some or all of those commits if need be.

I think it would be good to get this backported to 0.6 as well, at least the first commit. It seems that there's only one tiny conflict, so it shouldn't be a big problem.

Example build: http://mortenpi.eu/julia/julia-repl/index.html

mortenpi committed:
Use julia-repl in documentation

With the highlighting updates in Documenter v0.10.1 it is now possible
to have specialized highlighting for REPL blocks via the julia-repl
attribute on code blocks. Doctests already use this, so this makes the
highlighting for non-doctest REPL blocks consistent.

mortenpi committed:
Use at-repl for versioninfo() output

mortenpi committed:
Enable highlighting in the manual

A couple of code examples were not highlighted for some reason.

mortenpi committed:
Remove highlighting from REQUIRE code

These blocks are not Julia code, but special REQUIRE file syntax
instead.

mortenpi committed:
Enable bash highlighting

Also add a newline to the end of the file.

mortenpi committed:
Edit manual/conversion-and-promotion.md

The REPL example is basically exactly as above, and does not add
anything to the latter bit about the actual convert methods.

mortenpi commented on doc/src/devdocs/backtraces.md:
````diff
@@ -18,18 +18,8 @@ No matter the error, we will always need to know what version of Julia you are r
-  LIBM: libopenlibm
-  LLVM: libLLVM-3.3
+```@repl
+versioninfo()
````
This change would also require a small bugfix in Documenter (JuliaDocs/Documenter.jl#484).

tkelman commented on doc/src/devdocs/subarrays.md:
````diff
@@ -152,7 +152,7 @@ we can define dispatch directly on `SubArray{T,N,A,I,true}` without any intermed
 happens to be uniform:
 
-```julia
+```julia-repl
````
doctest candidate?

mortenpi replied:
Ideally we'd doctest a lot more in this file, but we'd need a way to run ordinary code within doctests. A named `@example` block sharing scope with doctests could be a way maybe, but Documenter doesn't support that at the moment afaict.

tkelman commented on doc/src/manual/arrays.md:
````diff
@@ -182,7 +182,7 @@ julia> sum(1/n^2 for n=1:1000)
 are needed to separate the generator from subsequent arguments:
 
-```julia
+```julia-repl
````
doctest candidate? in case the error message changes I suppose

mortenpi replied:
Documenter can't handle syntax errors as far as I can tell (at the moment, should probably be fixed).

tkelman replied:
Ah my mistake then. Is there a tracking issue?

mortenpi replied:
There is now: JuliaDocs/Documenter.jl#487 Also, the same applies to the invalid UTF-8 example in `manual/strings.md`.

tkelman commented on doc/src/manual/conversion-and-promotion.md:
````diff
@@ -66,20 +66,20 @@ julia> typeof(ans)
 doesn't know how to perform the requested conversion:
 
-```julia
+```julia-repl
````
doctest too for the error message?

KristofferC commented:
I'm not sure I understand why both `julia` and `julia-repl` need to exist? What is the difference?

tkelman commented on doc/src/manual/packages.md:
````diff
@@ -106,7 +106,7 @@ that the `Distributions` package should be installed since it is required but no
 
-```julia
+```julia-repl
 $ echo UTF16 >> ~/.julia/v0.6/REQUIRE
````
we may want to change this example to a non-deprecated package

tkelman commented on doc/src/manual/running-external-programs.md:
````diff
@@ -49,7 +49,7 @@ true
 More generally, you can use [`open()`](@ref) to read from or write to an external command.
 
-```julia
+```julia-repl
````
doctestable?

tkelman commented on doc/src/manual/strings.md:
````diff
@@ -808,7 +808,7 @@ The Unicode escape `\u2200` is encoded in UTF-8 as the three bytes 226, 136, 128
 a regular string literal, you will get a syntax error:
 
-```julia
+```julia-repl
````
could doctest the error

tkelman commented on doc/src/manual/variables-and-scoping.md:
````diff
@@ -174,7 +174,7 @@ julia> x
 when it would change the semantics is (currently) a syntax error:
 
-```julia
+```julia-repl
````
doctest for the error?

tkelman commented on doc/src/manual/variables-and-scoping.md:
````diff
@@ -332,7 +332,7 @@ The reason to allow *modifying local* variables of parent scopes in nested funct
 have a private state, for instance the `state` variable in the following example:
 
-```julia
+```julia-repl
````
this, and the last in this file, should also be testable?

tkelman commented on doc/src/stdlib/test.md:
````diff
@@ -26,7 +26,7 @@ Base.Test.@test_throws
 For example, suppose we want to check our new function `foo(x)` works as expected:
 
-```julia
+```julia-repl
````
could we but the series in this file in a common named doctest set?

mortenpi replied:
In principle this should be doable, but it's a bit more work. I'd leave this for another day.

ararslan commented:
@KristofferC The `julia` definition highlights code blocks as programs. That is, every keyword is highlighted. When used as we do throughout the docs, to simulate a REPL session, the output is also highlighted as though it were code. The difference with `julia-repl` is that it highlights the `julia>` prompt and all code that's part of the input, but applies no highlighting to the output. Basically it better simulates what a REPL session looks like.

mortenpi committed:
Fix line numbers in doctests

mortenpi committed:
Fix a doctest in base/path.jl

mortenpi committed:
Use a non-deprecated package in an example

UTF16.jl has been deprecated for a while. Use SHA.jl instead as the
example.

mortenpi commented on doc/src/manual/methods.md:
````diff
@@ -163,7 +163,7 @@ f (generic function with 2 methods)
 of those methods are, use the [`methods()`](@ref) function:
 
-```julia
+```julia-repl
````
This can't be a doctest at the moment since Documenter thinks that the first line of output is a comment that should be ignored (JuliaDocs/Documenter.jl#483).

mortenpi commented:
Rebased and all the doctests seem to be passing now. I can squash some of the commits (e.g. all the "Enable doctests in ..." ones) if need be, but I was trying to keep distinct changes in separate commits.

KristofferC commented:
Great with so many more doctests!

tkelman commented on base/regex.jl:
````diff
@@ -75,7 +75,7 @@ after the ending quote, to change its behaviour:
 For example, this regex has all three flags enabled:
 
-```julia
+```julia-repl
````
did I comment on this one or miss it, seems doctestable?

mortenpi committed:
Enable a doctest in base/regex.jl

StefanKarpinski commented:
We could probably add some smarts to a single `julia` highlight mode so that it understands when a code block looks like a REPL interaction and does the right thing, but I think having these be different modes is ok too – it provides useful semantic information and it's not all that onerous to mark the two kinds of blocks differently.

mortenpi commented:
> We could probably add some smarts to a single julia highlight mode

Someone has to implement that in highlight.js though. As you point out, there are pros and cons to both approaches, so I'm actually neutral on the matter. For some data: other languages in highlight.js seem to have [separate `lang-repl` modes](https://github.com/isagalaev/highlight.js/pull/1511#discussion_r114703162), and both Atom's [language-julia](https://github.com/JuliaEditorSupport/atom-language-julia/blob/master/grammars/julia-console.cson) and [Pygments](https://bitbucket.org/birkenfeld/pygments-main/src/7941677dc77d4f2bf0bbd6140ade85a9454b8b80/pygments/lexers/julia.py?at=default&fileviewer=file-view-default#julia.py-284) have a separate grammar/lexer for Julia's REPL syntax -- although that might be just because it's easier to implement it this way.

However, in any case, switching back to a single `julia` mode would be simple and non-breaking, so we could easily go with `julia-repl` for now and change back later if we want, I think.

KristofferC commented:
Barring objections I plan to merge this later today.

ararslan commented:
It would be good to squash some of the commits first; 23 is a lot of separate commits for this.

mortenpi committed:
Enable doctests the the manual

Switching to julia-repl revealed several code blocks that could
trivially be turned into doctests.

mortenpi commented:
Done, down to 11 now. I squashed the similar ones. I left the doctest ones under `base/` separate, so it would be a bit easier with any potential merge conflicts, should they get backported.

ararslan commented:
Perfect. Thanks, @mortenpi!
